core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+
+                remoteLogMetadataManager.updateRemoteLogSegmentMetadata(rlsmAfterSegmentCopy).get()
+
+                copiedOffsetOption = Some(endOffset)
+                log.updateHighestOffsetInRemoteStorage(endOffset)
+                info(s"Copied $logFileName to remote storage with segment-id: ${rlsmAfterSegmentCopy.remoteLogSegmentId()}")
+              }
+            }
+          } else {
+            debug(s"Skipping copying segments, current read-offset:$copiedOffset, and LSO:$lso")
+          }
+        }
+      } catch {
+        case ex: Exception =>
+          if (!isCancelled()) {
+            error(s"Error occurred while copying log segments of partition: $tpId", ex)
+          }
+      }
+    }
+
+    private def toPathIfExists(file: File): Path = {
+      if (file.exists()) file.toPath else null
+    }
+
+    override def run(): Unit = {
+      if (isCancelled())
+        return
+
+      try {
+        if (isLeader()) {
+          // Copy log segments to remote storage
+          handleCopyLogSegmentsToRemote()
+        }
+      } catch {
+        case ex: InterruptedException =>
+          if (!isCancelled()) {
+            warn(s"Current thread for topic-partition-id $tpId is interrupted, this task won't be rescheduled. " +
+              s"Reason: ${ex.getMessage}")
+          }
+        case ex: Exception =>
+          if (!isCancelled()) {
+            warn(s"Current task for topic-partition $tpId received error but it will be scheduled. " +
+              s"Reason: ${ex.getMessage}")
+          }
+      }
+    }
+
+    override def toString: String = {
+      this.getClass.toString + s"[$tpId]"
+    }
+  }
+
+  def findHighestRemoteOffset(topicIdPartition: TopicIdPartition): Long = {
+    var offset: Optional[lang.Long] = Optional.empty()
+    fetchLog(topicIdPartition.topicPartition()).foreach { log =>
+      log.leaderEpochCache.foreach(cache => {
+        var epoch = cache.latestEpoch
+        while (!offset.isPresent && epoch.isPresent) {
+          offset = remoteLogMetadataManager.highestOffsetForEpoch(topicIdPartition, epoch.getAsInt)
+          epoch = cache.previousEpoch(epoch.getAsInt)
+        }
+      })
+    }
+    offset.orElse(-1L)
+  }
+
+
+  private def doHandleLeaderOrFollowerPartitions(topicPartition: TopicIdPartition,
+                                                 convertToLeaderOrFollower: RLMTask => Unit): Unit = {
+    val rlmTaskWithFuture = leaderOrFollowerTasks.computeIfAbsent(topicPartition, (tp: TopicIdPartition) => {
+      val task = new RLMTask(tp)
+      // set this upfront when it is getting initialized instead of doing it after scheduling.
+      convertToLeaderOrFollower(task)
+      info(s"Created a new task: $task and getting scheduled")
+      val future = rlmScheduledThreadPool.scheduleWithFixedDelay(task, 0, delayInMs, TimeUnit.MILLISECONDS)
+      RLMTaskWithFuture(task, future)
+    })
+    convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)
```

Comment:
Will there be race condition when the topic partition is new inserted, so it'll set `convertToLeaderOrFollower` once before scheduling. And then the scheduled thread could be run immediately, later, run this line `convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)`, so it reset the leader epoch if it got updated. 

Could we add a null condition check for `rlmTaskWithFuture` here? that is:
```
if (rlmTaskWithFuture != null) {
  // not new created task, should convert it here
 convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)
}
```

Reply:
There is no race condition here. `rlmTaskWithFuture` will never be null as it will be created if it does not exist in `leaderOrFollowerTasks` map. It is expected that leader or follower switches can happen resulting leader epoch updates and `RLMTask` takes care of handling those updates.